README.md
[![Build Status](https://github.com/axetroy/anti-redirect/workflows/build/badge.svg)](https://github.com/axetroy/anti-redirect/actions)
[![DeepScan grade](https://deepscan.io/api/teams/5773/projects/7595/branches/79869/badge/grade.svg)](https://deepscan.io/dashboard#view=project&tid=5773&pid=7595&bid=79869)
[![996.icu](https://img.shields.io/badge/link-996.icu-red.svg)](https://996.icu)
[![LICENSE](https://img.shields.io/badge/license-Anti%20996-blue.svg)](https://github.com/996icu/996.ICU/blob/master/LICENSE)

### GM 脚本，反重定向

去除各搜索引擎/常用网站的重定向

> 注意事项：
>
> 重定向一般有两种目的
>
> 1. 追踪用户打开了哪些 URL
> 2. 在用户跳转到站外之前进行确认地址，防止打开不明的页面

### 反馈地址

> 反馈最好能带上出问题的网页地址

- https://github.com/axetroy/anti-redirect/issues/new/choose
- https://github.com/axetroy/anti-redirect/issues/new/choose
- https://github.com/axetroy/anti-redirect/issues/new/choose

### 如果这能够帮助到你, 不妨点个 star, 你的支持就是我更新的动力

[点击从 Github 安装](https://github.com/axetroy/anti-redirect/raw/gh-pages/anti-redirect.user.js)

[点击从 GreasyFork 安装](https://greasyfork.org/scripts/11915-anti-redirect-typescript/code/anti-redirect.user.js)

[点击从 CDN 安装(国内用户)](https://cdn.jsdelivr.net/gh/axetroy/anti-redirect@gh-pages/anti-redirect.user.js)

### 工作原理

1.  根据 URL 上暴露出来的跳转链接，正则匹配提取真实的地址，例如知乎，Google
2.  如果 A 标签的内容为真实的地址，则替换，例如百度贴吧
3.  逐一发送请求，获取真实的地址，例如百度搜索
4.  根据请求特殊页面，这个特殊页面没有重定向地址，然后覆盖当前页，例如百度搜索，搜狗搜索
5.  覆盖原本的链接点击事件，比如 qq 邮箱

### 更新日志

[https://github.com/axetroy/anti-redirect/blob/master/CHANGELOG.md](https://github.com/axetroy/anti-redirect/blob/master/CHANGELOG.md)

### 支持的站点

- [x] 知乎
- [x] 知乎专栏
- [x] 知乎日报
- [x] Google 搜索
- [x] Google 文档
- [x] Google Play
- [x] Google Gmail
- [x] Google Youtube
- [x] Steam
- [x] 360 搜索
- [x] 新浪微博
- [x] Twitter
- [x] 搜狗搜索
- [x] 百度搜索
- [x] 百度视频
- [x] 百度学术
- [x] 百度贴吧
- [x] 掘金
- [x] QQ 邮箱
- [x] Mozilla
- [x] 简书
- [x] 豆瓣
- [x] Pocket
- [x] DogeDoge
- [x] 秘迹
- [x] CSDN
- [x] 开源中国
- [x] 印象笔记
- [x] 标志情报局
- [x] 爱发电
- [x] 51 CTO
- [x] InfoQ
- [x] Gitee
- [x] 少数派

<details><summary>更多</summary>

- [x] 51.ruyo.net

</details>

### 我想支持更多的站点

点击这个[链接](https://github.com/axetroy/anti-redirect/issues/new)，提交 issues，说出你想要支持的站点

### 贡献代码

需要通过 NodeJs 把 TypeScript 编译成 javascript

```bash
git clone https://github.com/axetroy/anti-redirect.git

cd ./anti-redirect

npm install
npm run watch
```

### 开源许可

The [Anti 996 License](https://github.com/axetroy/anti-redirect/blob/master/LICENSE)

请仔细阅读开源许可。

简而言之： 如果你正在 996，或者你的公司/单位正在 996 ，那么请不要安装这个脚本！


CHANGELOG.md
## v2.21.1 (2023-07-21)

### Bugs fixed:

- youtube([`ea804d1`](https://github.com/axetroy/anti-redirect/commit/ea804d125cb9de9fa3fccf290d4a87d1b8fb0905)) (by Axetroy)

## v2.21.0 (2023-05-29)

### New feature:

- 增加少数派的支持, close #620([`27a53ed`](https://github.com/axetroy/anti-redirect/commit/27a53ed5f4f03985638fa0bff419e1b8aa4b6b7d)) (by Axetroy)
- 支持 csdn 自定义域名, close #623([`30ae45f`](https://github.com/axetroy/anti-redirect/commit/30ae45fec6c3d7a4d91c05d61ae0152fe77f510e)) (by Axetroy)
- 支持 gitee.com, close #641([`0dacbf2`](https://github.com/axetroy/anti-redirect/commit/0dacbf2883ddd5bed3f4b1bd11304df343bb34a6)) (by Axetroy)
- add xie.infoq.cn([`d4df9b2`](https://github.com/axetroy/anti-redirect/commit/d4df9b24bdbef29d17eaa20e7ebf555f0d75771a)) (by Axetroy)

### Bugs fixed:

- format([`bf5fb2a`](https://github.com/axetroy/anti-redirect/commit/bf5fb2a4be1ed2694eb87a7158086d6b163ff346)) (by Axetroy)
- lint([`a059c15`](https://github.com/axetroy/anti-redirect/commit/a059c15f54a3d1f33a736388943bd350c132ff32)) (by Axetroy)
- lint([`eaf94e8`](https://github.com/axetroy/anti-redirect/commit/eaf94e81adf8997aa29863c708c4a21644f66704)) (by Axetroy)

## v2.20.0 (2022-10-03)

### New feature:

- 支持 51 CTO. close #566([`ddfa551`](https://github.com/axetroy/anti-redirect/commit/ddfa5510e01841d94a3fb0f20f687fb743a5100c)) (by Axetroy)
- 支持爱发电. close #598([`35d0b1b`](https://github.com/axetroy/anti-redirect/commit/35d0b1bd392ceb6505cf5a60721d7c8627098449)) (by Axetroy)

### Bugs fixed:

- 修复 Google 的其他地方站点. close #599([`2a89351`](https://github.com/axetroy/anti-redirect/commit/2a8935123243f89951ad40545444f8741eea33ee)) (by Axetroy)

## v2.19.6 (2022-05-02)

### Bugs fixed:

- google([`c64381e`](https://github.com/axetroy/anti-redirect/commit/c64381e9f6bc7fee108e3daf3af5743ca5b0278c)) (by Axetroy)
- google([`ddfcb55`](https://github.com/axetroy/anti-redirect/commit/ddfcb55ad05df07561ff477a454b3f7e7bc951af)) (by Axetroy)

## v2.19.5 (2022-04-10)

### Bugs fixed:

- 提升百度的稳定性([`0efa80f`](https://github.com/axetroy/anti-redirect/commit/0efa80fde242eeaf092cc9d69b547dd3a8e78397)) (by Axetroy)

## v2.19.4 (2022-04-10)

### New feature:

- 支持标志情报局. close #481([`84e7231`](https://github.com/axetroy/anti-redirect/commit/84e72315d65fbf75e4b84d56718fb8f81621bdba)) (by Axetroy)

## v2.19.3 (2022-04-10)

## v2.19.2 (2022-04-10)

### New feature:

- 支持 juejin.cn([`e26c6a2`](https://github.com/axetroy/anti-redirect/commit/e26c6a25834c7638dffe0db3fc27cb2e51ee7880)) (by Axetroy)

### Bugs fixed:

- twitter. close #565([`b43b527`](https://github.com/axetroy/anti-redirect/commit/b43b52715b64107bd71799d344e2ead2296c396a)) (by Axetroy)

## v2.19.0 (2021-03-22)

### New feature:

- 支持 oschina([`e2e15c7`](https://github.com/axetroy/anti-redirect/commit/e2e15c77a90011b8dd301f506a2765e3260b9c90)) (by axetroy)

## v2.18.7 (2021-03-09)

## v2.18.6 (2021-03-09)

### Bugs fixed:

- 修复 CSDN 的 TOC 重定向会把 hash 地址的跳转错误([`c1a2c2e`](https://github.com/axetroy/anti-redirect/commit/c1a2c2e8377d2a486ef0146d6e8c4a0f3d84c198)) (by axetroy)

## v2.18.5 (2021-01-20)

### Bugs fixed:

- 修复简书有些链接没有去除重定向 close #366([`b5cb3c7`](https://github.com/axetroy/anti-redirect/commit/b5cb3c7eca7ad8fb79ec9c678710e8a205860db4)) (by axetroy)

## v2.18.4 (2021-01-02)

## v2.18.3 (2020-12-31)

## v2.18.2 (2020-12-31)

## v2.18.0 (2020-12-25)

### New feature:

- 支持印象笔记. close #344([`ef4a5e2`](https://github.com/axetroy/anti-redirect/commit/ef4a5e2ebb251f37cfb02c17b92e5ff5f109930e)) (by axetroy)

## v2.17.0 (2020-12-25)

### New feature:

- 支持 CSDN. close #350([`814b162`](https://github.com/axetroy/anti-redirect/commit/814b16297f3a4c8588c13644da06cd2eb1b6d4fd)) (by axetroy)

## v2.16.2 (2020-10-25)

### Bugs fixed:

- #277([`4c0908d`](https://github.com/axetroy/anti-redirect/commit/4c0908d0dcd2254af5f18585e86ba9088a3f3bbe)) (by axetroy)

## v2.16.0 (2020-09-23)

### New feature:

- 添加秘迹的支持 close #251([`6a10c3c`](https://github.com/axetroy/anti-redirect/commit/6a10c3cb40e351534d333cbc051f79887eb422da)) (by axetroy)

### Bugs fixed:

- 修复在搜索结果中打不开的百度文库 close #212([`f845d24`](https://github.com/axetroy/anti-redirect/commit/f845d24d006c9db92829c9938976a52ee10ab8f8)) (by axetroy)

## v2.15.2 (2020-06-07)

### New feature:

- 支持 youtube channel 页 (#243)([`5bfd7e9`](https://github.com/axetroy/anti-redirect/commit/5bfd7e942214cc34b3660995a9a057146666ac54)) (by 人造电子小猫咪)

## v2.15.1 (2020-04-10)

### Bugs fixed:

- 修复 360 搜索. close #204([`f16b81c`](https://github.com/axetroy/anti-redirect/commit/f16b81cdd451527e69859aa65917214154737611)) (by axetroy)

## v2.15.0 (2020-03-24)

### New feature:

- 支持 steam. close #48([`0256a46`](https://github.com/axetroy/anti-redirect/commit/0256a4667ccbdf38f3fcb4cf10cb70500a09d73e)) (by axetroy)

### Bugs fixed:

- 修复简书部分页面没有去除重定向 close #199([`3d8c78b`](https://github.com/axetroy/anti-redirect/commit/3d8c78b7d95d6d69c78277ef2dea6e6253e57e39)) (by axetroy)

## v2.14.5 (2020-02-26)

### Bugs fixed:

- 适配 Google 的新追踪策略. close #176([`5cb1da0`](https://github.com/axetroy/anti-redirect/commit/5cb1da06d1d6337c922363a539c50984ed17b556)) (by axetroy)

## v2.14.4 (2020-02-11)

## v2.14.3 (2020-02-11)

### New feature:

- 部分网站的去除重定向请求，不再携带 cookie([`279d80c`](https://github.com/axetroy/anti-redirect/commit/279d80c39fe1859b6d9a61e1eb09f6c6e9f9ca7f)) (by axetroy)

## v2.14.2 (2019-12-11)

### Bugs fixed:

- 简书([`56503b4`](https://github.com/axetroy/anti-redirect/commit/56503b45aac07708258a289595adf6248e7f31a1)) (by axetroy)

## v2.14.1 (2019-11-26)

### Bugs fixed:

- update([`ea15955`](https://github.com/axetroy/anti-redirect/commit/ea15955d55445e57c905c74fb98815aac48e53cb)) (by axetroy)

## v2.14.0 (2019-11-26)

### New feature:

- support ruyo. close #155([`4f3bded`](https://github.com/axetroy/anti-redirect/commit/4f3bdeda6a3f7f4da0c02c383f72372f63c8b812)) (by axetroy)

## v2.13.1 (2019-09-28)

## v2.13.0 (2019-09-23)

### New feature:


index.ts
import { App } from "@/app";
import { RuyoProvider } from "@/sites/51.ruyo.net";
import { MozillaProvider } from "@/sites/addons.mozilla.org";
import { YinXiangProvider } from "@/sites/app.yinxiang.com";
import { CSDNProvider } from "@/sites/blog.csdn.net";
import { OSChinaProvider } from "@/sites/oschina.com";
import { ZhihuDailyProvider } from "@/sites/daily.zhihu.com";
import { GoogleDocsProvider } from "@/sites/docs.google.com";
import { PocketProvider } from "@/sites/getpocket.com";
import { GmailProvider } from "@/sites/gmail.google.com";
import { JuejinProvider } from "@/sites/juejin.com";
import { QQMailProvider } from "@/sites/mail.qq.com";
import { MiJiProvider } from "@/sites/mijisou.com";
import { GooglePlayProvider } from "@/sites/play.google.com";
import { SteamProvider } from "@/sites/steamcommunity.com";
import { TiebaProvider } from "@/sites/tieba.baidu.com";
import { TwitterProvider } from "@/sites/twitter.com";
import { BaiduVideoProvider } from "@/sites/video.baidu.com";
import { WeboProvider } from "@/sites/weibo.com";
import { BaiduProvider } from "@/sites/www.baidu.com";
import { DogeDogeProvider } from "@/sites/www.dogedoge.com";
import { DouBanProvider } from "@/sites/www.douban.com";
import { GoogleProvider } from "@/sites/www.google.com";
import { JianShuProvider } from "@/sites/www.jianshu.com";
import { SoProvider } from "@/sites/www.so.com";
import { SoGouProvider } from "@/sites/www.sogou.com";
import { YoutubeProvider } from "@/sites/www.youtube.com";
import { ZhihuProvider } from "@/sites/www.zhihu.com";
import { BaiduXueshuProvider } from "@/sites/xueshu.baidu.com";
import { ZhihuZhuanlanProvider } from "@/sites/zhuanlan.zhihu.com";
import { LogonewsProvider } from "@/sites/www.logonews.cn";
import { AfDianNetProvider } from "@/sites/afadian.net";
import { Blog51CTO } from "@/sites/blog.51cto.com";
import { InfoQProvider } from "@/sites/infoq.cn";
import { GiteeProvider } from "@/sites/gitee.com";
import { SSPaiProvider } from "@/sites/sspai.com";
import http from "gm-http";

const app = new App();
const isDebug: boolean = process.env.NODE_ENV !== "production";

http.setConfig({ debug: isDebug });

app
  .setConfig({ isDebug })
  .registerProvider([
    {
      // 测试地址: https://www.zhihu.com/question/25258775
      name: "知乎",
      test: /www\.zhihu\.com/,
      provider: ZhihuProvider,
    },
    {
      // 测试地址: https://zhuanlan.zhihu.com/p/20549978
      name: "知乎专栏",
      test: /zhuanlan\.zhihu\.com/,
      provider: ZhihuZhuanlanProvider,
    },
    {
      // 测试地址:
      name: "知乎日报",
      test: /daily\.zhihu\.com/,
      provider: ZhihuDailyProvider,
    },
    {
      name: "Google搜索",
      test: /\w+\.google\./,
      provider: GoogleProvider,
    },
    {
      // 测试地址: https://docs.google.com/spreadsheets/d/1TFcEXMcKrwoIAECIVyBU0GPoSmRqZ7A0VBvqeKYVSww/htmlview
      name: "Google Docs",
      test: /docs\.google\.com/,
      provider: GoogleDocsProvider,
    },
    {
      name: "Gmail",
      test: /mail\.google\.com/,
      provider: GmailProvider,
    },
    {
      // 测试地址: https://play.google.com/store/movies/details/%E7%A7%BB%E5%8B%95%E8%BF%B7%E5%AE%AE_%E6%AD%BB%E4%BA%A1%E8%A7%A3%E8%97%A5?id=YNy7gRqwtMk
      name: "Google Play",
      test: /play\.google\.com/,
      provider: GooglePlayProvider,
    },
    {
      // 测试地址: https://www.youtube.com/watch?v=XTXSRRSv1bY
      name: "Google Youtube",
      test: /www\.youtube\.com/,
      provider: YoutubeProvider,
    },
    {
      // 测试地址: https://www.so.com/s?ie=utf-8&fr=none&src=360sou_newhome&q=chrome
      name: "360搜索",
      test: /www\.so\.com/,
      provider: SoProvider,
    },
    {
      name: "新浪微博",
      test: /\.weibo\.com/,
      provider: WeboProvider,
    },
    // 测试: https://twitter.com/ftium4/status/1512815116810522631
    {
      name: "Twitter",
      test: /twitter\.com/,
      provider: TwitterProvider,
    },
    {
      // 测试: http://www.sogou.com/web?query=chrome&_asf=www.sogou.com&_ast=&w=01019900&p=40040100&ie=utf8&from=index-nologin&s_from=index&sut=1527&sst0=1504347367611&lkt=0%2C0%2C0&sugsuv=00091651B48CA45F593B61A29B131405&sugtime=1504347367611
      name: "搜狗搜索",
      test: /www\.sogou\.com/,
      provider: SoGouProvider,
    },
    {
      // 测试: https://www.baidu.com/s?wd=chrome&rsv_spt=1&rsv_iqid=0xcb136237000ed40e&issp=1&f=8&rsv_bp=0&rsv_idx=2&ie=utf-8&tn=baidulocal&rsv_enter=1&rsv_sug3=7&rsv_sug1=7&rsv_sug7=101&rsv_sug2=0&inputT=813&rsv_sug4=989&timestamp=1504349229266&rn=50&vf_bl=1
      name: "百度搜索",
      test: /www\.baidu\.com/,
      provider: BaiduProvider,
    },
    {
      // 测试: https://www.baidu.com/s?wd=chrome&pn=20&oq=chrome&tn=baiduhome_pg&ie=utf-8&usm=3&rsv_idx=2&rsv_pq=e043900d0000752d&rsv_t=6bb0UqEwp2Tle6TAMBDlU3Wg%2BSxoqvvOhZKyQgM%2BVQP8Gc54QZLhcDcj62eGfNG75aq5&rsv_page=1
      name: "百度视频",
      test: /v\.baidu\.com/,
      provider: BaiduVideoProvider,
    },
    {
      // 测试: http://xueshu.baidu.com/s?wd=paperuri%3A%28ae4d6b5da05eca552dab05aeefb966e6%29&ie=utf-8&filter=sc_long_sign&sc_ks_para=q%3D%E2%80%9C%E4%BA%92%E8%81%94%E7%BD%91%2B%E5%81%A5%E5%BA%B7%E7%AE%A1%E7%90%86%E2%80%9D%E6%A8%A1%E5%BC%8F%E6%8E%A2%E8%AE%A8%E5%8F%8A%E5%85%B6%E5%BA%94%E7%94%A8&tn=SE_baiduxueshu_c1gjeupa
      name: "百度学术",
      test: /xueshu\.baidu\.com/,
      provider: BaiduXueshuProvider,
    },
    {
      // 测试地址: http://tieba.baidu.com/p/5300844180
      name: "百度贴吧",
      test: /tieba\.baidu\.com/,
      provider: TiebaProvider,
    },
    {
      // 测试地址: https://juejin.im/entry/59ac8fa551882524241a8802?utm_source=gold_browser_extension
      name: "掘金",
      test: /juejin\.(im|cn)/,
      provider: JuejinProvider,
    },
    {
      name: "QQ邮箱",
      test: /mail\.qq\.com/,
      provider: QQMailProvider,
    },
    {
      // 测试地址: https://addons.mozilla.org/zh-CN/firefox/addon/evernote-web-clipper/
      name: "Mozilla",
      test: /addons\.mozilla\.org/,
      provider: MozillaProvider,
    },
    {
      // 测试地址: https://www.jianshu.com/p/979776ca44b8
      // https://www.jianshu.com/p/fc8abc65bbb2
      name: "简书",
      test: /www\.jianshu\.com/,
      provider: JianShuProvider,
    },
    {
      // 测试地址: https://www.douban.com/doulist/240962/
      // 测试地址: https://www.douban.com/search?cat=1002&q=%E9%BB%91%E9%95%9C
      name: "豆瓣",
      test: /douban\.com/,
      provider: DouBanProvider,
    },
    {
      // 测试地址: https://getpocket.com/a/recommended/
      // 需要登陆
      name: "Pocket",
      test: /getpocket\.com/,
      provider: PocketProvider,
    },
    {
      // 测试地址: https://www.dogedoge.com/results?q=chrome
      name: "DogeDoge",
      test: /www\.dogedoge\.com/,
      provider: DogeDogeProvider,
    },
    {
      // 测试地址: https://51.ruyo.net/15053.html
      name: "Ruyo",
      test: /51\.ruyo\.net/,
      provider: RuyoProvider,
    },
    {
      // 测试地址: https://steamcommunity.com/sharedfiles/filedetails/?id=1311535531
      name: "Steam",
      test: /steamcommunity\.com/,
      provider: SteamProvider,
    },
    {
      // 测试地址: https://mijisou.com/?q=chrome&category_general=on&time_range=&language=zh-CN&pageno=1
      name: "秘迹",
      test: /mijisou\.com/,
      provider: MiJiProvider,
    },
    {
      // 测试地址: https://github.com/axetroy/anti-redirect/issues/350
      name: "CSDN",
      test: /blog\.csdn\.net/,
      provider: CSDNProvider,
    },
    {
      // 测试地址：https://my.oschina.net/chipo/blog/3067672
      name: "OS China",
      test: /oschina\.net/,
      provider: OSChinaProvider,
    },
    {
      // 测试地址: https://github.com/axetroy/anti-redirect/issues/350
      name: "印象笔记",
      test: /app\.yinxiang\.com/,
      provider: YinXiangProvider,
    },
    {
      // 测试地址: https://www.logonews.cn/2021073002420141.html
      name: "标志情报局",
      test: /www\.logonews\.cn/,
      provider: LogonewsProvider,
    },
    {
      // 测试地址: https://afdian.net/a/xiaofanEric
      name: "爱发电",
      test: /afdian\.net/,
      provider: AfDianNetProvider,
    },
    {
      // 测试地址: https://blog.51cto.com/u_11512826/2068421
      name: "51CTO博客",
      test: /blog\.51cto\.com/,
      provider: Blog51CTO,
    },
    {
      // 测试地址: https://xie.infoq.cn/link?target=https%3A%2F%2Fwww.finclip.com%2F%3Fchannel%3Dinfoqseo
      name: 'InfoQ',
      test: /infoq\.cn/,
      provider: InfoQProvider
    },
    {
      // 测试地址: https://gitee.com/Tencent/ncnn
      name: 'Gitee',
      test: /gitee.com/,
      provider: GiteeProvider
    },
    {
      // 测试地址: https://sspai.com/post/77499
      name: '少数派',
      test: /sspai\.com/,
      provider: SSPaiProvider
    }
  ])
  .bootstrap();
